cmd/serve/s3/backend.go

```diff
@@ -0,0 +1,497 @@
+
+	in, err := file.Open(os.O_RDONLY)
+	if err != nil {
+		return nil, gofakes3.ErrInternal
+	}
+	defer func() {
+		// If an error occurs, the caller may not have access to Object.Body in order to close it:
+		if err != nil {
+			_ = in.Close()
+		}
+	}()
+
+	var rdr io.ReadCloser = in
+	rnge, err := rangeRequest.Range(size)
+	if err != nil {
+		return nil, err
+	}
+
+	if rnge != nil {
+		if _, err := in.Seek(rnge.Start, io.SeekStart); err != nil {
+			return nil, err
+		}
+		rdr = limitReadCloser(rdr, in.Close, rnge.Length)
+	}
+
+	meta := map[string]string{
+		"Last-Modified": node.ModTime().Format(timeFormat),
+		"Content-Type":  fs.MimeType(context.Background(), fobj),
+	}
+
+	if val, ok := tmpMetaStorage.Load(fp); ok {
+		metaMap := val.(map[string]string)
+		for k, v := range metaMap {
+			meta[k] = v
+		}
+	}
+
+	return &gofakes3.Object{
+		Name:     gofakes3.URLEncode(objectName),
+		Hash:     hash,
+		Metadata: meta,
+		Size:     size,
+		Range:    rnge,
+		Contents: rdr,
+	}, nil
+}
+
+// TouchObject creates or updates meta on specified object.
+func (db *s3Backend) TouchObject(fp string, meta map[string]string) (result gofakes3.PutObjectResult, err error) {
+
+	_, err = db.fs.Stat(fp)
+	if err == vfs.ENOENT {
+		f, err := db.fs.Create(fp)
+		if err != nil {
+			return result, err
+		}
+		_ = f.Close()
+		return db.TouchObject(fp, meta)
+	} else if err != nil {
+		return result, err
+	}
+
+	_, err = db.fs.Stat(fp)
+	if err != nil {
+		return result, err
+	}
+
+	tmpMetaStorage.Store(fp, meta)
+
+	if val, ok := meta["X-Amz-Meta-Mtime"]; ok {
+		ti, err := swift.FloatStringToTime(val)
+		if err == nil {
+			return result, db.fs.Chtimes(fp, ti, ti)
+		}
+		// ignore error since the file is successfully created
+	}
+
+	if val, ok := meta["mtime"]; ok {
+		ti, err := swift.FloatStringToTime(val)
+		if err == nil {
+			return result, db.fs.Chtimes(fp, ti, ti)
+		}
+		// ignore error since the file is successfully created
+	}
+
+	return result, nil
+}
+
+// PutObject creates or overwrites the object with the given name.
+func (db *s3Backend) PutObject(
+	bucketName, objectName string,
+	meta map[string]string,
+	input io.Reader, size int64,
+) (result gofakes3.PutObjectResult, err error) {
+
+	_, err = db.fs.Stat(bucketName)
+	if err != nil {
+		return result, gofakes3.BucketNotFound(bucketName)
+	}
+
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	fp := path.Join(bucketName, objectName)
+	objectDir := path.Dir(fp)
+	// _, err = db.fs.Stat(objectDir)
+	// if err == vfs.ENOENT {
+	// 	fs.Errorf(objectDir, "PutObject failed: path not found")
+	// 	return result, gofakes3.KeyNotFound(objectName)
+	// }
+
+	if objectDir != "." {
+		if err := mkdirRecursive(objectDir, db.fs); err != nil {
+			return result, err
+		}
+	}
+
+	if size == 0 {
+		// maybe a touch operation
+		return db.TouchObject(fp, meta)
+	}
+
+	f, err := db.fs.Create(fp)
+	if err != nil {
+		return result, err
+	}
+
+	hasher := md5.New()
+	w := io.MultiWriter(f, hasher)
```

Comment:
This md5 hash object gets initialized, and then it gets a copy of all the bytes from the input. However hasher.Sum(nil) is never called, and the md5 results seem to be never used at all. There is a report that PUT operations use up memory equal to twice the input filesize, and never releases it, eventually resulting in an OOM condition. Could the md5 hasher be holding on to the memory and never releasing it because .Sum() is never called?

And if nothing is ever done with the md5sum, why not just remove it entirely?